samcli/lib/logs/formatter.py

```diff
@@ -0,0 +1,220 @@
+                return event
+
+        Parameters
+        ----------
+        colored : samcli.lib.utils.colors.Colored
+            Used to add color to the string when pretty printing. Colors are useful only when pretty printing on a
+            Terminal. To turn off coloring, set the appropriate property when instantiating the
+            ``samcli.lib.utils.colors.Colored`` class.
+
+        formatter_chain : list of formatter methods
+
+        """
+
+        self.colored = colored
+        self.formatter_chain = formatter_chain or []
+
+        # At end of the chain, pretty print the Event object as string.
+        self.formatter_chain.append(LogsFormatter._pretty_print_event)
+
+    def do_format(self, event_iterable):
+        """
+        Formats the given CloudWatch Logs Event dictionary as necessary and returns an iterable that will
+        return the formatted string. This can be used to parse and format the events based on context
+        ie. In Lambda Function logs, a formatter may wish to color the "ERROR" keywords red,
+        or highlight a filter keyword separately etc.
+
+        This method takes an iterable as input and returns an iterable. It does not immediately format the event.
+        Instead, it sets up the formatter chain appropriately and returns the iterable. Actual formatting happens
+        only when the iterable is used by the caller.
+
+        Parameters
+        ----------
+        event_iterable : iterable of samcli.lib.logs.event.LogEvent
+            Iterable that returns an object containing information about each log event.
+
+        Returns
+        -------
+        iterable of string
+            Iterable that returns a formatted event as a string.
+        """
+
+        for operation in self.formatter_chain:
+
+            # Make sure the operation has access to certain basic objects like colored
+            partial_op = functools.partial(operation, colored=self.colored)
+            event_iterable = map(partial_op, event_iterable)
+
+        return event_iterable
+
+    @staticmethod
+    def _pretty_print_event(event, colored):
+        """
+        Basic formatter to convert an event object to string
+
+        Parameters
+        ----------
+        event
+        colored
+
+        Returns
+        -------
+
+        """
+        event.timestamp = colored.yellow(event.timestamp)
+        event.log_stream_name = colored.cyan(event.log_stream_name)
+
+        return ' '.join([event.log_stream_name, event.timestamp, event.message])
+
+
+class LambdaLogMsgFormatters(object):
+    """
+    Format logs printed by AWS Lambda functions.
+
+    This class is a collection of static methods that can be used within a formatter chain.
+    """
+
+    @staticmethod
+    def colorize_errors(event, colored):
+        """
+
+        Parameters
+        ----------
+        message
+
+        Returns
+        -------
+
+        """
+
+        nodejs_crash_msg = "Process exited before completing request"
```

Comment:
Should these be an input into the function? Seems a little strange to have these hardcoded into the actually function. It's not the end of the world like this but just caught my eye having some specific cases defined here.

Reply:
It can be, but at some level you do need to define them within a function. Each of these formatters are self-contained and potentially replaceable. So I kept everything isolated inside.